plog-kafka/src/main/java/com/airbnb/plog/kafka/KafkaHandler.java

```diff
@@ -13,29 +13,31 @@
 import lombok.RequiredArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
 
-import java.io.ByteArrayOutputStream;
-import java.util.concurrent.atomic.AtomicLong;
-
 import javax.crypto.Cipher;
 import javax.crypto.spec.SecretKeySpec;
+import java.io.ByteArrayOutputStream;
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicLong;
 
 @RequiredArgsConstructor
 @Slf4j
 public final class KafkaHandler extends SimpleChannelInboundHandler<Message> implements Handler {
     private final String defaultTopic;
     private final boolean propagate;
-    private final Producer<byte[], byte[]> producer;
+    private final Producer<Integer, byte[]> producer;
```

Comment:
I see.  The benefit of keeping byte[] for key is we can extend plog to accept customized key through the tag field in the future. Therefore, the application can overwrite the key with more meaningful semantics. If the key type is integer, then we need another refactor for this.

Another point is Object.hashcode works perfectly/better in this semantics. User could set the same key for multiple messages. The default partitioner will evenly distribute this key across partitions (because we hash the object, not hash actual bytes). So the application does not need to worry about the key distribution and adding randomness into the key partition. In addition, if they really want to control the partition algorithm based on key. Then they probably need to use a customized partitioner.

IMO, using object.hashcode for byte[] is a "feature" not a "bug", which provides a neat way to decouple the message_key with partition_key :)